tsehelper to generate json for system extensions

goproslowyo opened the pull request:
This commit adds the tsehelper tool, facilitating the creation of a mapping schema for talos system extensions from a specified container registry. The tool utilizes environment variables for configuration and offers flexibility through flags. It introduces idempotent operations, cache refresh, and various output format flags. This enhancement streamlines talos system extensions management for users.

* Idempotent Operation: ensures it only fetches versions not already cached
* Configurable: set environment variables to configure the tool's behavior, such as specifying the container registry and cache directory.
* Output Format Options: flags enable users to customize the output format, including minimal JSON output, cache purging, and trimming of registry, SHA256, and tag details.

budimanjojo commented:
Thank you so much! Looking at it now and I got one question. Is the cache file being updated if there's a change, like there's a new version added? I can't find it in the code but maybe it's just my bad go skill.

goproslowyo commented:
> Is the cache file being updated if there's a change, like there's a new version added? I can't find it in the code but maybe it's just my bad go skill.
> 
> EDIT: Oh I think there's. missed the `getMissingVersions` function.

Yep, take a look at the [README.md](https://github.com/budimanjojo/talhelper/pull/266/files#diff-849722b169760dab8249570776d5ad317d29b2d7d376660d26c9536621c66672R12) and let me know if you've got any questions!

I just added two more flags as well, `-version v1.5.5` and `-onlyVersions` to return only version numbers.

budimanjojo commented on hack/tsehelper/main.go:
```diff
@@ -0,0 +1,114 @@
+				// Sort the cache
+				sort.Sort(&tags)
+				// Write the cache file
+				writeCache(&tags)
```
The error is not being checked?

budimanjojo commented on hack/tsehelper/main.go:
```diff
@@ -0,0 +1,114 @@
+		// Sort the cache
+		sort.Sort(&tags)
+		// Save the cache file
+		writeCache(&tags)
```
Same as above

budimanjojo commented on hack/tsehelper/cache.go:
```diff
@@ -0,0 +1,90 @@
+	path := getCachePath()
+	os.Remove(path)
+	// Write blank cache file
+	writeCache(&TalosVersionTags{})
```
Here too

budimanjojo commented:
Everything is looking fine, let me know when you're done with the commits and I'll merge it. The only gripe is the regex part, but maybe that's because the crane library is not providing any exported type for the `Descriptor`? Thank you so much!

goproslowyo commented:
> Everything is looking fine, let me know when you're done with the commits and I'll merge it. The only gripe is the regex part, but maybe that's because the crane library is not providing any exported type for the `Descriptor`? Thank you so much!

I am done making changes, merge at will!

Here's a link to help make sense of the regex with a ton of example strings: https://regex101.com/r/sjl2E6/1

If you have some ideas on how you'd like to see if change let me know and I can make later commits. I originally was parsing with strings.Split but it quickly became too messy and the regex instantly cleaned up everything and made it extremely fast versus string matching.

budimanjojo commented:
Thank you!

budimanjojo commented:
@goproslowyo I think there's something being added by the upstream. The workflow failed: https://github.com/budimanjojo/talhelper/actions/runs/7333886972

I don't know what's happening but there's some changes to the cache file in latest version:
```json
        {
            "version": "v1.6.1-1-gd0879ba",
            "systemExtensions": [
                "ghcr.io/siderolabs/amdgpu-firmware:20231111@sha256:4f0b95b466ba29e78a688114369e38a358a6f5ca7399b29cca67aed9aa8d2b62:",
                "  author: Sidero Labs",
                "  description: |",
                "    This system extension provides AMDGPU firmware binaries.",
                "ghcr.io/siderolabs/amd-ucode:20231111@sha256:ebf5b6e47b58044398c1d83bf7c78f671fb841907411ce4bc19f45043c60f83f:",
                "  author: Sidero Labs",
                "  description: |",
                "    This system extension provides AMD microcode binaries.",
```
I don't really have a clue on how to fix this :cry:

budimanjojo commented:
maybe the we need to filter those strings in the `processTarArchive` function?

budimanjojo commented:
I got it fixed by adding a `|| !regexp.Match(line)`  inside the https://github.com/budimanjojo/talhelper/blob/d5f316659d9e9f3bf8a27afdadb54fed14bf4872/hack/tsehelper/tar_archive.go#L37

But, maybe there's another way to handle this more beautifully?

goproslowyo commented:
Hey @budimanjojo, just arose and made some coffee. Will take a look. I think the regex is brittle and this code from a friend might be a bit more robust (and maybe faster too):

```golang
package main

import (
	"fmt"
	"strings"
)

func main() {
	sysExtensionURL := "ghcr.io/siderolabs/btrfs:v1.6.0@sha256:c1a7dbc0475c1794fdbd9599758953def54984548fade10fc094d53d5dafe9f0"
	urlSplit := strings.Split(sysExtensionURL, "/")
	domain := urlSplit[0]
	org := urlSplit[1]
	imageSplit := strings.Split(urlSplit[2], ":")
	image := imageSplit[0]
	version := strings.Split(imageSplit[1], "@")[0]
	sha256Hash := strings.Split(sysExtensionURL, ":")[2]

	fmt.Printf("Domain: %s, org: %s, image: %s, version: %s, hash: %s\n", domain, org, image, version, sha256Hash)
}
Domain: ghcr.io, org: siderolabs, image: btrfs, version: v1.6.0, hash: c1a7dbc0475c1794fdbd9599758953def54984548fade10fc094d53d5dafe9f0
```

goproslowyo commented:
@budimanjojo Tracked it down to [this PR](https://github.com/siderolabs/extensions/commit/57503ccd3bbba510401dbde67340b71eee0adfee) that broke the formatting of the file.

goproslowyo commented:
Opened https://github.com/budimanjojo/talhelper/pull/279 as a fix.
